exporters/geneva/src/exporter.cc

```diff
@@ -0,0 +1,359 @@
+  }
+  for (const auto &kv : attributes) {
+    auto attr_value = AttributeValueToString(kv.second);
+    SerializeString(buffer_non_histogram_, bufferIndex, attr_value);
+  }
+  // length zero for auto-pilot
+  SerializeInt<uint16_t>(buffer_non_histogram_, bufferIndex, 0);
+
+  // get final size of payload to be added in front of buffer
+  uint16_t body_length = bufferIndex - kBinaryHeaderSize;
+
+  // Add rest of the fields in front of buffer
+  bufferIndex = 0;
+
+  // event_type
+  SerializeInt<uint16_t>(buffer_non_histogram_, bufferIndex,
+                         static_cast<uint16_t>(event_type));
+
+  // body length
+  SerializeInt<uint16_t>(buffer_non_histogram_, bufferIndex,
+                         static_cast<uint16_t>(body_length));
+
+  // count of dimensions.
+  SerializeInt<u_int16_t>(buffer_non_histogram_, bufferIndex,
+                          static_cast<uint16_t>(attributes.size()));
+
+  // reserverd word (2 bytes)
+  SerializeInt<uint16_t>(buffer_non_histogram_, bufferIndex, 0);
+
+  // reserved word (4 bytes)
+  SerializeInt<uint32_t>(buffer_non_histogram_, bufferIndex, 0);
+
+  // timestamp utc (8 bytes)
+  auto windows_ticks = UnixTimeToWindowsTicks(
+      std::chrono::duration_cast<std::chrono::duration<std::uint64_t>>(
+          ts.time_since_epoch())
+          .count());
+
+  SerializeInt<uint64_t>(buffer_non_histogram_, bufferIndex, windows_ticks);
+  if (event_type == MetricsEventType::ULongMetric) {
+    SerializeInt<uint64_t>(buffer_non_histogram_, bufferIndex,
+                           static_cast<uint64_t>(nostd::get<long>(value)));
+  } else {
+    SerializeInt<uint64_t>(
+        buffer_non_histogram_, bufferIndex,
+        *(reinterpret_cast<const uint64_t *>(&(nostd::get<double>(value)))));
+  }
+  return body_length;
+}
+
+size_t Exporter::SerializeHistogramMetrics(
+    sdk::metrics::AggregationType agg_type, MetricsEventType event_type,
+    uint64_t count, const sdk::metrics::ValueType &sum,
+    const sdk::metrics::ValueType &min, const sdk::metrics::ValueType &max,
+    const std::list<double> &boundaries,
+    const std::vector<uint64_t> &counts, common::SystemTimestamp ts,
+    std::string metric_name, const sdk::metrics::PointAttributes &attributes) {
+
+  auto bufferIndex = buffer_index_histogram_;
+  // metric name
+  SerializeString(buffer_histogram_, bufferIndex, metric_name);
+
+  // dimentions - name
+  for (const auto &kv : attributes) {
+    if (kv.first.size() > kMaxDimensionNameSize) {
+      LOG_WARN("Dimension name limit overflow: %s", kv.first.c_str());
+      continue;
+    }
+    SerializeString(buffer_histogram_, bufferIndex, kv.first);
+  }
+
+  // dimentions - value
+  for (const auto &kv : attributes) {
+    auto attr_value = AttributeValueToString(kv.second);
+    SerializeString(buffer_histogram_, bufferIndex, attr_value);
+  }
+
+  // two bytes padding for auto-pilot
+  SerializeInt<uint16_t>(buffer_histogram_, bufferIndex, 0);
+
+  // version - set as 0
```

Comment:
Is there a document on all these fields and their length?

Reply:
The Kaitai DSL (https://github.com/open-telemetry/opentelemetry-cpp-contrib/pull/206/files#diff-065a936660be59a10e24c6f4d67a78d98a9f9fe7058f364b233fa2ec00afe771) is the best possible documentation for now. Will see if the ingestion agent has any protocol structure diagram available, else will try to document it in separate PR.